Implement Kubernetes2 adapter (based on new APA template model).

guptasu opened the pull request:
**What this PR does / why we need it**:

NOTE: The cache*.go files are completely unchanged.

**One concerns with this version of kubernetes2 compared to old code**.
Compared to the existing code inside adapter/kubernetes,
to me it felt we do not need any init-time locking on the controller
object with this new model where a new build is instantiated for each
adapter config. But I might be completely wrong. PTAL at the existing
code inside [BuildAttributeGenerator](https://github.com/istio/istio/blob/master/mixer/adapter/kubernetes/kubernetes.go#L217)
and the new code inside [Build](https://github.com/istio/istio/compare/master...guptasu:kub5?expand=1#diff-68714f93ebcea2e7e31119ca8fdda59aR191)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
none
```

istio-testing commented:
@guptasu: Adding do-not-merge/release-note-label-needed because the release note process has not been followed.

<details>

One of the following labels is required "release-note", "release-note-action-required", or "release-note-none".
Please see: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed.

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

mandarjog commented:
If cache code is not new, just import it from the old place. Kubernetes2 will be renamed to kubernetes shortly anyway.

guptasu commented:
Unfortunately functions are private in cache*.go. Is it easy to ignore
those 2 files during review than me editing the existing cache code to make
it accessible from other dir. Let me know if you really prefer importing, I
can do that.

On Dec 4, 2017 7:20 PM, "mandarjog" <notifications@github.com> wrote:

> If cache code is not new, just import it from the old place. Kubernetes2
> will be renamed to kubernetes shortly anyway.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/1982#issuecomment-349183017>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AK6xE6p1bpShMcKZ8c73kZ1z_bLGox3Uks5s9LaQgaJpZM4Q1pYd>
> .
>

guptasu commented:
@douglas-reid and @mjog , I know you guys are at the conference and busy, but in case you have some free time, can you please take a look at this PR. My concern is about, me not creating a Singleton controller anymore inside the `Build` function. Let me know if it is completely wrong.

mjog commented:
dude, you have the wrong @mjog


On 5 December 2017 17:50:51 Sunny Gupta <notifications@github.com> wrote:

> @douglas-reid and @mjog , I know you guys are at the conference and busy, 
> but in case you have some free time, can you please take a look at this PR. 
> My concern is about, me not creating a Singleton controller anymore inside 
> the `Build` function. Let me know if it is completely wrong.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/istio/istio/pull/1982#issuecomment-349212379

elevran commented:
coverage tests are failing (exit code 1), possibly due to bazel vs go build env.
```sh
go test -coverprofile=/tmp/coverage/istio.io-istio-mixer-adapter-list.txt istio.io/istio/mixer/adapter/list
mixer/adapter/kubernetes2/kubernetes.go:39:2: no Go files in /go/src/istio.io/istio/mixer/adapter/kubernetes2/config
mixer/adapter/kubernetes2/kubernetes.go:40:2: no Go files in /go/src/istio.io/istio/mixer/adapter/kubernetes2/template
```
I'm assuming that there are generated files (`.pb.go`) but they aren't created in the directory itself (but in the bazel generated directory) and thus not visible to go tool chain during the run.
It is an error to import packages with no go files.

jmuk commented:
please run `bin/regenerate_files.py` and bring those generated `pb.go` file and `gen.go` files in this PR. It should also change toplevel `generated_files` file.

BTW, why those new templates are in `mixer/adapter/kubernetes2/template` rather than `mixer/template`?  Also missing changes on `mixer/template/BUILD`?

jmuk commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,441 @@
+	paramsProto := b.adapterConfig
+	stopChan := make(chan struct{})
+	var pods cacheController
+	if b.needsCacheInit {
```
by the way -- `sync.Once` would be simpler and safer, I think (see https://golang.org/pkg/sync/#example_Once)

guptasu replied:
PTAL at the change now since it is always initcache

jmuk commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,441 @@
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	paramsProto := b.adapterConfig
+	stopChan := make(chan struct{})
+	var pods cacheController
```
This would be wrong. Second run of `Build` will leave this `pod` to be `nil`.
Let's add a test case of building multiple handlers from a single builder.

guptasu replied:
Awesome catch. thanks.. added a test and removed the `needsCacheInit = false` code. WDYT ?

jmuk replied:
It's even better to remove `needsCacheInit` itself if it always init the cache

guptasu replied:
yup.. D

guptasu commented:
@jmuk answer for your question 
_BTW, why those new templates are in mixer/adapter/kubernetes2/template rather than mixer/template? Also missing changes on mixer/template/BUILD?_

* The APA template is adapter specific and therefore it is inside kubernetes adapter directory.
* There is a separate PR that is still not committed which is needed for me to add this new template to mixer/template/BUILD

jmuk commented:
(sorry, accidentally hit 'close and comment' button...)

No problem if `mixer/template/BUILD` change will come as another PR. I was worrying because the config storage (and other modules of mixs) will not recognize those new templates until it's within mixer/template list.
I'm also wondering if this dependency (mixer/template -> mixer/adapter/kubernetes/template) is detected as a cycle dependency, but I'm not sure.

guptasu commented:
@jmuk if there is a cyclic dep, then it is a separate issue and should be solved.. But I think  we should be fine, will see.

jmuk commented on mixer/adapter/kubernetes2/cache.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	controllerImpl struct {
+		clientset     kubernetes.Interface
```
`clientset` is only used for initializing SharedInformer, seems unnecessary to store here.

jmuk commented on mixer/adapter/kubernetes2/cache.go:
```diff
@@ -0,0 +1,266 @@
+	namespace := "" // todo: address unparam linter issue
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
```
You can simply pass as `clientset.Pods(namespace)` here, since the first parameter for the shared informer is actually an interface and `Pods()` satisfies it.

douglas-reid replied:
I suspect a large chunk of these comments are related to updates to the client-go libraries and their contracts since the original time of writing `cache.go`. 

We don't have to revisit in this PR, but we should probably file several issues related to updating (and validating) this cache impl.

jmuk replied:
filed https://github.com/istio/istio/issues/2046
By the way, my comment here is actually wrong, `clientset.Pods(namespace)` can't be used directly. I believe other parts can be simplified.

jmuk commented on mixer/adapter/kubernetes2/cache.go:
```diff
@@ -0,0 +1,266 @@
+	}
+	c.env.ScheduleDaemon(func() {
+		c.pods.Run(stop)
+		c.env.Logger().Infof("pod cache started")
```
IIRC, `Run` will block until it ends (i.e. `stop` is closed). So this logging message would be wrong.

jmuk commented on mixer/adapter/kubernetes2/cache.go:
```diff
@@ -0,0 +1,266 @@
+		c.pods.Run(stop)
+		c.env.Logger().Infof("pod cache started")
+	})
+	<-stop
```
If it synchronously wait for stop, I don't think we should schedule `c.pods.Run` in another goroutine, simply invoke here directly.

jmuk commented on mixer/adapter/kubernetes2/cache.go:
```diff
@@ -0,0 +1,266 @@
+		},
+		&v1.Pod{},
+		refreshDuration,
+		cache.Indexers{},
```
You can create the index of IP address to the pod here. This way, you don't have to keep `ipPodMap` anymore, and you don't have to have the event handler below.

jmuk commented on mixer/adapter/kubernetes2/cache_test.go:
```diff
@@ -0,0 +1,281 @@
+	return f
+}
+
+type fakeInformer struct {
```
Rather than creating fake informer, it is better to have fake clientset (i.e. `k8s.io/kubernetes/fake`).

istio-testing commented:
@guptasu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | d5a2debd1712f4f506829ae2ca96451b6af5acd1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1982/istio-presubmit/1007/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

guptasu commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,436 @@
+// GetInfo returns the Info associated with this adapter implementation.
+func GetInfo() adapter.Info {
+	return adapter.Info{
+		Name:        "KubernetesAttributeGenerator",
```
changed to kubernetesenv after talking to @geeknoid . We can revisit this name if need be.

guptasu commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,436 @@
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		env.Logger().Infof("Cache sync successful.")
+	}
+	kg := &handler{
```
D

istio-testing commented:
@guptasu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | a4e2cc9b2d67df61fcd0e6bfd869494bb835f060 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1982/istio-presubmit/1026/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

jmuk commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,436 @@
+
+	// cache invalidation
+	// TODO: determine a reasonable default
+	defaultRefreshPeriod = 5 * time.Minute
```
That is not correct actually, this value is for fully refreshing (re-building) the local cache. Every this period, the local cache internally abandons the whole cached data, gets the full data from k8s apiserver, and reconstruct the cached data again. Specifying `0` means this whole rebuilding is not going to happen.

On the other hand, cache can watch updates through its `Watch` method, and this works regardless of the refresh period. Also the cache can reconnect this watching periodically. So the cache is unlikely to be out-of-sync without refreshing.

jmuk replied:
ah, but again, since this is in cache.go file, I think it is okay to leave unchanged. We can revisit this later.

jmuk commented:
/lgtm

jmuk commented:
btw I'm writing some changes on cache.go for what I commented, as https://github.com/jmuk/istio/commit/31a5bc85c542e2303fe6110799c163a266132a96. Will catch up with this once this is merged.

guptasu commented:
I just rebased and pushed.. Can you please re lgtm and approve

guptasu commented:
@jmuk can you please re-lgtm and re-approve

douglas-reid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,409 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"context"
```
linter issue (or should be).

douglas-reid commented on mixer/adapter/kubernetes2/template/template.proto:
```diff
@@ -0,0 +1,89 @@
+    istio.mixer.v1.template.IPAddress origin_ip = 6;
+}
+
+message OutputTemplate {
```
we need significant documentation around each of these fields. This existed before. Not adding it here is a regression.

guptasu replied:
For comments I was looking at the wrong place before. I looked at the config in testdata. You are referring to documentation in config.. got it.. thanks

updated..

douglas-reid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,409 @@
+	"context"
+
+	"istio.io/istio/mixer/adapter/kubernetes2/config"
+	kubernetes_apa_tmpl "istio.io/istio/mixer/adapter/kubernetes2/template"
```
nit: make the import name much shorter. something like: `ktmpl` or `k8stmpl`.

guptasu replied:
D

douglas-reid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,409 @@
+var _ kubernetes_apa_tmpl.HandlerBuilder = &builder{}
+
+/////////////////////////////// Config-time methods //////////////////////////////
+//
```
what are these extra lines of `//` needed for?

guptasu replied:
Copied from one the test where I was dividing the config code from runtime code.. I guess, it is not that important.

douglas-reid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,409 @@
+}
+
+/////////////////////////////// Request-time methods //////////////////////////////
+//
```
what are these extra lines of `//` needed for?

guptasu replied:
Removed

douglas-reid commented:
i think this PR is still missing the generated files for the new template.proto, no?

guptasu commented:
@douglas-reid Friendly ping

guptasu commented:
@jmuk @douglas-reid , can one of you please LGTM this PR if things look good. I have addressed last set of comments.

guptasu commented:
@douglas-reid @jmuk @geeknoid I have repurposed this PR to do a little more than before. In the last [4 commits (see just the delta here](https://github.com/istio/istio/pull/1982/files/f37174e9445c9b689cebd163adab72bfeba93293..a553043dbb92d4b3053a569583ad310e4d4d041b)) I have addressed @douglas-reid  comments + the following PTAL.
* added a testdata sample config for kubernetes2, 
* connected the template and the adapter to the Mixer machinery via the inventory, 
* removed old kubernetes from inventory and its associated testdata files.
* added generated files

This PR now officially cuts legacy kubernetes adapter from Mixer machinery. Next PR will be deletion of 
* legacy kubernetes adapter, 
* associated tests
* everything legacy (aspects etc.)

istio-testing commented:
@guptasu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | a553043dbb92d4b3053a569583ad310e4d4d041b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1982/e2e-smoke/657/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

guptasu commented:
/retest

istio-testing commented:
@guptasu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | a553043dbb92d4b3053a569583ad310e4d4d041b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1982/e2e-smoke/664/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@guptasu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 6faa53110a90acc816aa6824f04f03f569b392f1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/1982/e2e-smoke/681/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

geeknoid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,401 @@
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		env.Logger().Infof("Waiting for kubernetes cache sync...")
+	}
+	if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
```
We don't want to block handler creation on a potentially long running operation, especially one that can fail non-deterministically.  I think we could end up preventing config from being loaded and the handler started because the cache can't currently be refreshed.

geeknoid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,401 @@
+}
+
+func (h *handler) Close() error {
+	return nil
```
Don't you need to shut down the caching daemon that was started in Build()?

geeknoid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,401 @@
+	}
+	if app, found := p.Labels[params.PodLabelForService]; found {
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
```
What about when err != nil? We're not handling the failure case. Same for the code below. Should we at least log something?

geeknoid commented on mixer/adapter/kubernetes2/kubernetes.go:
```diff
@@ -0,0 +1,401 @@
+	if err != nil || config == nil {
+		return nil, fmt.Errorf("could not retrieve kubeconfig: %v", err)
+	}
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
```
Thoughout this source file, you don't need all these "if" checks around calling the logger. 

The only time you need to protect with an if like that, is if the actual call to the logger is expensive, such as if it requires calling fmt.Sprint before invoking the log method.

douglas-reid commented on install/kubernetes/helm/istio/charts/mixer/templates/istio-mixer.yaml:
```diff
@@ -190,6 +190,41 @@ spec:
+spec:
+  group: config.istio.io
+  names:
+    kind: kubernetes
```
i wonder if this will ever be confusing for operators. I don't have any great suggestions, however.

douglas-reid commented on install/kubernetes/helm/istio/charts/mixer/templates/istio-mixer.yaml:
```diff
@@ -767,3 +802,55 @@ spec:
+  name: handler
+  namespace: {{$istio_namespace}}
+spec:
+  # when running from mixer root, use the following config after adding a
```
this comment probably isn't useful for people installing via helm.

guptasu replied:
D

douglas-reid commented on mixer/adapter/inventory.go:
```diff
@@ -15,15 +15,10 @@
 package adapter
```
do we need this file at all now?

guptasu replied:
In the next PR, I will delete this. Removing it now has cascading effects which I was trying to avoid in this PR

douglas-reid commented on mixer/adapter/kubernetes2/template/template.proto:
````diff
@@ -0,0 +1,134 @@
+//     destination.serviceAccount: $out.destination_service_account_name
+// ```
+message Template {
+    string source_uid = 1;
````
can we add comments here about the expected formats of this data ?

guptasu replied:
D

guptasu commented:
@douglas-reid , @geeknoid friendly ping
